Formula/synergy-core.rb

```diff
@@ -0,0 +1,274 @@
+class SynergyCore < Formula
+  desc "Synergy, the keyboard and mouse sharing tool"
+  homepage "https://symless.com/synergy"
+  url "https://github.com/symless/synergy-core/archive/refs/tags/1.14.3.5-stable.tar.gz"
+  sha256 "9259fc8aa02610121932cd5887c4d878beb08b2f7190e834c0584d53b36f5f9b"
+
+  # The synergy-core/LICENSE file contains the following preamble:
+  #    This program is released under the GPL with the additional exemption
+  #    that compiling, linking, and/or using OpenSSL is allowed.
+  # This preamble is followed by the text of the GPL-2.0.
+  #
+  # The list of SPDX license exceptions does not contain an exception corresponding
+  # to synergy-core's special OpenSSL exception. The openvpn-openssl-exception
+  # exception is very similar, but it is not the same, so it would be inappropriate
+  # to use it here.
+  #
+  # Unfortunately, writing an exception like this:
+  #     with: :cannot_represent
+  # causes `brew audit` to raise the following error:
+  #     Formula synergy-core contains invalid or deprecated SPDX license exceptions: [:cannot_represent].
+  # As a result, it appears we cannot mention the exception in the `license` definition.
+  license "GPL-2.0-only"
```

Comment:
If it's not a know exception that might just mean that it's not a valid exception.

Reply:
The original copyright holder can release their work under whatever terms they see fit. There's not really such a thing as an invalid exception, if the person granting the exception is the copyright holder. Anything that can be expressed in words can be a license so it's not surprising that something like SPDX can never be comprehensive.

Reply:
Okay, let me rephrase that. Maybe if the exception is not know that is because it makes the license incompatible with open source software.

Reply:
That could be a concern in some cases. To be clear, the exception here only gives people _more_ rights than they would have under the GPLv2. If the GPLv2 is free software and open source, then this license (call it the "Synergy license"), which is strictly more free than GPLv2, is also free source and open source.

That said, perhaps what you're getting at is that there may be an issue of whether a program licensed under the Synergy license can be _combined_ with a program licensed _only_ under the normal GPLv2. That could actually be a problem if certain conditions are met, but it's not a novel problem. There are various free software licenses that have the property that work licensed under their terms cannot be combined with work licensed under GPLv2. The key question is whether we'd be distributing such a combination -- and I don't know the answer to that, because I haven't looked closely at everything in the synergy-core repository.

However, the potential issue isn't in any way unique to synergy-core; any given brew formula could potentially combine free software programs with incompatible licenses. Whether `brew` can legally distribute the bottles built from those formulae would depend on a close reading of potentially all of the licenses of all of the direct or indirect dependencies of the project. Is this something that `brew` currently attempts to audit in a systematic way across its many formulae? Since we have rudimentary license information for each formula, some analysis is possible in an automated way, but it would have major limitations since a lot of the disputes would center on the exact wordings of the licenses, not so much the image of the dependency graph.

Reply:
I don't want this pull request to get sidetracked, so let me try to bring this back to what I think is relevant to this formula.

The [Formula Cookbook](https://docs.brew.sh/Formula-Cookbook) says that the test for inclusion in brew is whether the software has a "[Debian Free Software Guidelines license](https://wiki.debian.org/DFSGLicenses)" ("DFSG license").

The question of whether `synergy-core` can be distributed when combined with _other_ free software projects is not itself part of the test for inclusion in `brew`, and indeed, many things in `brew` might have potential trouble with that test: for example, [according to the Free Software Foundation](https://www.gnu.org/licenses/rms-why-gplv3.en.html) ("FSF"), anything licensed under "GPL-2.0-only" (~230 formulae in `homebrew-core`) cannot be distributed when "combine[d]" with something released under "GPL-3.0" (~771 formulae) or "GPL-3.0-only" (~147 formulae). However, the FSF goes on to assert, on the same page, that they don't think this incompatibility is a big problem because the incompatible programs can be, they say, distributed "side by side" so long as they aren't "combine[d]". The question of whther brew's bottling mechanism is consistent with the licenses of all projects in brew is an interesting thing to analyze and audit, but it should be a separately tracked issue, not specifically related to this `synergy-core` formula.

For _this_ formula, the question relevant to `brew` is, as mentioned, whether Synergy is licensed under a DFSG license. There are two sub-questions here:
- First, if Synergy is licensed under the license stated in the license file at synergy-core/LICENSE (i.e., the GPLv2 with an exception granting people more rights than they would otherwise have under the GPLv2 and with no added restrictions (the "Synergy license")), is this a DGSF license? 
- Second, is Synergy _actually_ licensed under the Synergy license?

The answer to the first question is "yes", it is a DGSF license. The DGSF guidelines are found [here](https://www.debian.org/social_contract#guidelines). The GPLv2 is agreed to meet those guidelines so the only question is whether the addition of the Synergy exception -- allowing Synergy to be combined with OpenSSL -- causes a violation of the DGSF guidelines. And it seems that this Synergy exception does not affect whether the DGSF guidelines are met, because the Synergy license still allows free distribution, still allows redistribution of source code, still allows derived works, still does not mandate integrity of the author's source code, still does not discriminate against persons or groups, still does not discriminate against fields of endeavor, and still does not require parties to enter into any supplement licenses in order to gain rights to the work. So it is a DGSF-compatible license.

Indeed the Synergy license essentially the same license used by OpenVPN (included in `brew`). The only reason we can't use the symbol for the "[openvpn-openssl-exception](https://spdx.org/licenses/openvpn-openssl-exception.html)" exception is that the text of that exception reads as follows: "In addition, as a special exception, **OpenVPN Technologies, Inc. gives permission** to link the code of this program with the OpenSSL Library (or with modified versions of OpenSSL that use the same license as OpenSSL), and distribute linked combinations including the two." (Emphasis mine.) This symbol cannot be used for Synergy because "OpenVPN Technologies, Inc." does not have permission to grant rights to Synergy. Perhaps the SPDX symbol should be more general and not hardcore a specific company name, but since it does hardcode a company name, no one but OpenVPN can use that particular symbol. But even though we can't use the symbol, it is substantially the same license, which everyone agrees to agree is acceptable.

That leaves only the second question: is Synergy _actually_ licensed under the Synergy licensed? With any alleged free software program, this is always something that might be an issue. For example, somebody might redistribute a nonfree program and add a LICENSE file to the distribution that asserts that the program is licensed under GPLv2. Obviously, the mere inclusion of that LICENSE file would not cause the program to actually be licensed under GPLv2 because the person who added the file doesn't have the right to change the licensing terms unless authorized by the copyrght holders. Less nefariously, a person might accidentally combine two programs in a way that cannot be licensed under the original license of either program, causing whatever they say in the LICENSE file of the combined program to be inaccurate, without malicious intent. So just because a repository has a LICENSE file, it does not mean that the program is actually licensed under the license described in that file.

However, the `brew` project seems to generally take it at face value that LICENSE files can be trusted. We don't investigate every single commit to the repository going back the entire history of the project to make sure it's all consistent with the stated license. We generally trust the maintainers to accurately report the license, unless there's something obviously problematic going on. With `synergy-core`, there is nothing obviously problematic. The program has been around for almost 20 years and has always been free software. To be sure, Symless has often downplayed the free software nature of the software, because it wants to charge people money for things they don't necessarily have to pay for, but it has never attempted to make the software nonfree. It's a bit like the situation after Oracle acquired OpenOffice.org: Oracle didn't really promote the open source status of the software (leading to various forks, notably LibreOffice), but it also never did anything obviously improper. That's the same with Synergy. I can't claim that I've audited the entire history of the Synergy project, but if there is a problem, it would have to be a very subtle one, and that isn't normally something that blocks pull requests to `homebrew-core` absent evidence that there is a problem.

So, in my opinion, there isn't a licensing problem here. The most significant problem is with the SPDX symbols, because it strange to have a symbol (openvpn-openssl-exception) that can only ever be used by a single entity because the definition of the symbol hardcodes a particular company name. Perhaps we should consider making a pull request to the SPDX project, or at least creating an issue, to raise questions about this, but it shouldn't block this formula.

Reply:
The SPDX project appears to agree that GPL exceptions that allow combining the GPL program with non-GPL programs such as OpenSSL are acceptable. In spdx/license-list-XML#1442, we find the following:

> 1. Does the license substantially comply with one of the free/open content definitions?
>> Yes; it is an exception to the GPL and does not restrict use in any way.

I think this should be enough to remove any concerns about the Synergy license for inclusion in `brew`.

Reply:
> The SPDX project appears to agree that GPL exceptions that allow combining the GPL program with non-GPL programs such as OpenSSL are acceptable. In https://github.com/spdx/license-list-XML/issues/1442, we find the following:

No, the projects in question aren't GPL, they're GPL + exception, which makes them incompatible with the GPL. We've had to relicense code under the LGPL or GPL + exception to integrate it into those projects.

Reply:
No one is saying that "GPL + exception" is the same as "GPL". The question being discussed is **whether Synergy's "GPL + exception" is a free software license**. The answer is yes. The part of spd/license-list-XML#1442 I quoted was to demonstrate that **"GPL + linking exception" is a free software license**. It's not the same as the GPL, but it's still free software.

Reply:
I don't think the correct solution to brew not supporting this exception is to pretend there is no exception (except for a comment in the code that very few people will ever read). If brew has an allowlist for exceptions, it should be added there. The current statement here will mean we're lying about the license and I refuse to accept that.

Seeing how Debian ships a version of this it does seem to meet our requirements, but the current version is not correct.

Reply:
Okay, I'll find a solution that doesn't require having an inaccurate `license` statement.

Reply:
Here's the solution that I think is correct:

**I've changed the license statement to read `license :cannot_express`** because the Synergy license, although it is a free software license, cannot be expressed with brew's `license` statement.

This has precedent in `brew`. A few examples are as follows:

Formula/ngrep.rb:
```ruby
license :cannot_represent # Described as 'BSD with advertising' here: https://src.fedoraproject.org/rpms/ngrep/blob/rawhide/f/ngrep.spec#_8
```

Formula/libmagic.rb:
```ruby
# libmagic has a BSD-2-Clause-like license
license :cannot_represent
```

Formula/ispell.rb:
```ruby
license :cannot_represent # modified BSD license
```

The formulae for a few very well-known programs use `:cannot_represent` and don't even make any attempt to describe the license, including `krb5.rb` and `nmap.rb`.

I've also added a caveat telling the user where to find the `LICENSE` file. This information is not specific to `synergy-core` (indeed, brew retains the `LICENSE` files for all packages), but I think it should alleviate any concerns about the user not knowing what the license is. We can remove this if you wish.

****

Unfortunately, this new solution introduces another issue. Brew uses the [GitHub license API](https://docs.github.com/en/rest/licenses#get-the-license-for-a-repository) to determine the license of a project and if there is a mismatch between the `license` expression and the result of the API, brew raises an error. This causes a problem for `synergy-core` because the GitHub license API says the project is licensed under GPLv2. GitHub does not offer projects a way to override this API, so there is nothing the `synergy-core` people can do. I also looked into the open source component that GitHub uses to get license information (`licensee`) but the problem is **not** with that project. That project returns the following about Synergy's license:

```bash
$ bundle exec bin/licensee detect https://github.com/symless/synergy-core
License:        NOASSERTION
Matched files:  LICENSE
LICENSE:
  Content hash:  c917a967f2b8b39b2b038c7dd09c9d911e57d22d
  License:       NOASSERTION
  Closest non-matching licenses:
    GPL-2.0 similarity:  94.72%
    MPL-2.0 similarity:  42.31%
    EPL-2.0 similarity:  36.46%
```

So, the `licensee` program doesn't assert that synergy-core uses GPLv2; it just says that it uses a license that is _similar_ to GPLv2. However, GitHub's API ignores this detail and just says it uses GPLv2. This part of GitHub's API appears to be proprietary and undocumented. There is nowhere that I could find to read the code or submit a pull request. To workaround this issue, we need to add `synergy-core` to `audit_exceptions/permitted_formula_license_mismatches.json`, which is for formulae where the GitHub API returns the wrong result for the package being distributed.

Reply:
I'm still not entirely sure what prevents you from using a syntax like https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/gcc.rb#L16

Reply:
That was the first thing I tried before submitting the pull request. That syntax doesn't allow arbitrary strings for the license exception. Rather, `brew audit` downloads a JSON file from the internet and if the exception string isn't in that JSON file, it raises errors. For example, if we use `license "GPL-2.0-or-only" => { with: "synergy-core-openssl-exception" }`, these are the errors raised:

```
$ brew audit --strict --online --new-formula synergy
synergy-core:
  * Formula synergy-core contains non-standard SPDX licenses: ["GPL-2.0-or-only"].
    For a list of valid licenses check: https://spdx.org/licenses/
  * Formula synergy-core contains invalid or deprecated SPDX license exceptions: ["synergy-core-openssl-exception"].
    For a list of valid license exceptions check:
      https://spdx.org/licenses/exceptions-index.html
```

There's no current mechanism in brew's code to add to the contents of that JSON file, and the JSON URI referenced isn't under the control of the Homebrew project. If you're saying you want me to implement such a new whitelisting mechanism for strings not found in the remotely-hosted JSON file, we can do that, but I think what I've proposed in the current iteration of the pull request is what other formulae already do in this situation.